blockchain/blockchain.go

```diff
@@ -765,15 +770,47 @@ func revertHead(txn db.Transaction) error {
 	if err = txn.Delete(db.BlockHeaderNumbersByHash.Key(header.Hash.Marshal())); err != nil {
 		return err
 	}
+	if blockNumber != 0 { // Don't send block before genesis
+		var newHeader *core.Header
+		newHeader, err = blockHeaderByNumber(txn, blockNumber-1)
+		if err != nil {
+			return err
+		}
+		b.newHeads.Send(newHeader)
+	}
+
+	if err = removeTxsAndReceipts(txn, blockNumber, header.TransactionCount); err != nil {
+		return err
+	}
+
+	// remove state update
+	if err = txn.Delete(db.StateUpdatesByBlockNumber.Key(numBytes)); err != nil {
+		return err
+	}
+
+	// remove pending
+	if err = txn.Delete(db.Pending.Key()); err != nil {
+		return err
+	}
+
+	// update chain height
+	if blockNumber == 0 {
+		return txn.Delete(db.ChainHeight.Key())
+	}
 
+	heightBin := core.MarshalBlockNumber(blockNumber - 1)
+	return txn.Set(db.ChainHeight.Key(), heightBin)
+}
+
+func removeTxsAndReceipts(txn db.Transaction, blockNumber, numTxs uint64) error {
 	blockIDAndIndex := txAndReceiptDBKey{
 		Number: blockNumber,
 	}
 	// remove txs and receipts
-	for i := uint64(0); i < header.TransactionCount; i++ {
+	for i := uint64(0); i < numTxs; i++ {
 		blockIDAndIndex.Index = i
```

Comment:
I think we can make `blockIDAndIndex` declaration here entirely i.e.:
```
blockIDAndIndex := txAndReceiptDBKey{
	Number: blockNumber,
	Index:  i,
}
```

Reply:
Will compiler know that the final binary doesn't need to include write instructions for `blockIDAndIndex.Number`? 

It's totally a micro-optimization, but I'm curious nonetheless. I think we might as well go with the more efficient version rather than relying on the compiler.

Reply:
Short answer: you never know :) compilers are complex beasts and sometimes they apply all sorts of crazy unexpected optimizations.

In this particular example you are right current solution is more optimal than the one I proposed. Here is [output from godbolt](https://godbolt.org/z/xzej7aP41):
```
mov     rax, QWORD PTR [rbp-72]
mov     QWORD PTR [rbp-48], rax
mov     rax, QWORD PTR [rbp-8]
mov     QWORD PTR [rbp-40], rax
```
vs [assignment of just one field](https://godbolt.org/z/zP3x7fneY) 
```
mov     rax, QWORD PTR [rbp-8]
mov     QWORD PTR [rbp-40], rax
```

Reply:
Interesting, thanks. I think I'll leave it as it is.